### PR TITLE
[WIP]: Joining Code Validation and profile

### PR DESCRIPTION
**Pull request by zo-el**
#### Notes:
- **Profile:** I will update `hc-zomes` into the holochain org once it's approved.
- **Joining Code:** Since we need one joining code to send agents, for now, we are just signing the pub_key with the key itself.

### CODE REVIEW

File: tests/src/common.ts [hunk 1]
```diff
@@ -51,3 +52,29 @@ export const installation2agent: InstallAgentsHapps = [
   [[chatDna]],
   [[chatDna]],
 ]
+
+// this mem_proof is a signature of the holo_agent key signing its public key
+// The current holo_hosting pub key is `uhCAk7wGO_N3Lm9-OU7mDhyLSTI4WBxHQ9pq98TQbwaqmxJkqbmFW`
+const MEM_PROOF = Buffer.from("oTKk4HcEesGjQgj4aBfIfy/iQlot+LOYlbtj8p/Sn/NNYSRJ9xmbbysEO6wKvfP2y1VGdpj4Y3gNHT7HGVEuDw==", 'base64')
```
Comment (lucksus) [inline tests/src/common.ts:58]: Use this real joining-code instead:
```
eyJzaWduZWRfaGVhZGVyIjp7ImhlYWRlciI6eyJjb250ZW50Ijp7InR5cGUiOiJDcmVhdGUiLCJhdXRob3IiOnsidHlwZSI6IkJ1ZmZlciIsImRhdGEiOlsxMzIsMzIsMzYsMTI3LDYwLDE1Niw5MywyMDQsMTU2LDExOSwyMjcsMTQ2LDIzMiwxMTYsNywxOTAsMjAsMjI0LDEyMCw1Myw5OSwxNTAsNjUsOTMsMTksMTExLDE1MSwzMSw2MywxNTksMjUyLDAsMTkxLDIzMSwyMywxNjUsNjUsMTI1LDc3XX0sInRpbWVzdGFtcCI6WzE2MTU4OTc3NDcsODk5OTQxMDAzXSwiaGVhZGVyX3NlcSI6NCwicHJldl9oZWFkZXIiOnsidHlwZSI6IkJ1ZmZlciIsImRhdGEiOlsxMzIsNDEsMzYsMTA3LDIxMSw3MywxMTIsOTgsODksMTY3LDE5NywxOTgsMTM5LDIwNiw5LDU4LDEzNiwxNjMsMTQxLDkyLDE2LDU0LDE5OSwxNjEsMjMzLDIyOSw0LDI1MywyNDAsMzcsMjA0LDE0LDEzOCw1NCwzNiwyMjcsMTUyLDE3NiwxNDddfSwiZW50cnlfdHlwZSI6eyJBcHAiOnsiaWQiOjAsInpvbWVfaWQiOjAsInZpc2liaWxpdHkiOnsiUHVibGljIjpudWxsfX19LCJlbnRyeV9oYXNoIjp7InR5cGUiOiJCdWZmZXIiLCJkYXRhIjpbMTMyLDMzLDM2LDU3LDkwLDI1MiwxMjAsMTk3LDIyMywyNTQsMTY3LDIwMiwxNTMsMjQ1LDEsMTEyLDE4MywzOCwyNDgsMTk0LDgsMTYxLDIzNiwyMTksMjYsMTYwLDE5MSwxMDcsMTA3LDIwOCwyNDMsMjUzLDE3OSwyMTEsMjI5LDExNCwyMzUsNjAsMTM5XX19LCJoYXNoIjp7InR5cGUiOiJCdWZmZXIiLCJkYXRhIjpbMTMyLDQxLDM2LDI1NSw1OSwyMDcsMTYxLDEyNSw2NSwzMiwwLDY3LDY1LDM2LDE4OSwxMjcsMTU0LDE1Niw4NywxOTYsMTc4LDEyNyw5LDIzOCwxOCwxNjIsMjA4LDE2MywyMjgsMjIwLDI0Myw0OSwxOTAsMywxODcsODcsMTk3LDQ4LDIwOF19fSwic2lnbmF0dXJlIjp7InR5cGUiOiJCdWZmZXIiLCJkYXRhIjpbMywyMTksMzUsODUsMTIwLDIzNyw3OSwyMyw2NSw5OCwzNiw2MSw5Niw4OSwxMTksNjgsMTQ5LDE5Nyw4LDU2LDE0MiwyNDcsMTkwLDEyNiwxODYsNjUsOTksMzQsMTA5LDE2MywxNDQsMjgsMTE3LDE5NywxMjUsODcsMTAwLDEyNywxNTEsMjE5LDM1LDEwNywxNjUsMjE4LDM3LDU4LDE0NywxNDIsMjI0LDIyOSwxNjYsNjcsMjAxLDE0NCw2NSwxMDcsMjQwLDE4MiwxMTksMjM1LDE1NiwxOTAsMTEyLDZdfX0sImVudHJ5Ijp7IlByZXNlbnQiOnsiZW50cnlfdHlwZSI6IkFwcCIsImVudHJ5Ijp7InR5cGUiOiJCdWZmZXIiLCJkYXRhIjpbMTMwLDE2NCwxMTQsMTExLDEwOCwxMDEsMTY5LDEwMCwxMDEsMTE4LDEwMSwxMDgsMTExLDExMiwxMDEsMTE0LDE3NCwxMTQsMTAxLDk5LDExMSwxMTQsMTAwLDk1LDEwOCwxMTEsOTksOTcsMTE2LDExMSwxMTQsMTgwLDExMCwxMDUsOTksMTExLDEwOCw5NywxMTUsNDYsMTA4LDExNyw5OSwxMDcsNjQsMTA0LDEwMSwxMjEsNDYsOTksMTExLDEwOV19fX19
```

File: zomes/chat/src/validate.rs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+#[derive(Serialize, Deserialize, SerializedBytes, Debug )]
+pub struct JoiningCode(Signature);
+
+/// Validate joining code from the membrane_proof
+pub(crate) fn joining_code() -> ChatResult<()> {
+    let filter = QueryFilter::new();
+    let filter = filter.sequence_range(1..2);
+    let query_result: ElementVec = query(filter)?;
+    let holo_agent = AgentPubKey::try_from("uhCAk7wGO_N3Lm9-OU7mDhyLSTI4WBxHQ9pq98TQbwaqmxJkqbmFW").unwrap();
```
Comment (lucksus) [inline zomes/chat/src/validate.rs:14]: Use this pubkey that's used on the joining-code-server:
```
8420247f3c9c5dcc9c77e392e87407be14e078356396415d136f971f3f9ffc00bfe717a5417d4d
```

**Review by lucksus — COMMENTED** [review]
Ok, looks good so far. I've added real-world joining-code and public key as comments below.

After decoding that joining-code with base64, you'll find it to be a JSON object that can be deserialized into an `Element`. The entry of that element is the data about the user joining. In the header you'll find the signature and also the pubkey again, which should be the same as the hardcoded one. That is what should get checked in the `validate()`.

File: zomes/chat/src/validate.rs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+pub(crate) fn common_validatation(data: ValidateData) -> ExternResult<ValidateCallbackResult> {
+    let element = data.element;
+    let entry = element.into_inner().1;
+    let entry = match entry {
+        ElementEntry::Present(e) => e,
+        _ => return Ok(ValidateCallbackResult::Valid),
+    };
+    if let Entry::Agent(_) = entry {
+        return Ok(ValidateCallbackResult::Valid);
```
Comment (timotree3) [inline zomes/chat/src/validate.rs:46]: Shouldn't there be validation logic here that checks the joining code?

**Pull request closed without merging**